agr_literature_service/api/crud/search_crud.py

```diff
@@ -10,13 +12,90 @@
+    # convert string to Datetime int that is stored in Elastic search
+    # initial strings are in the format:- "2010-10-28T04:00:00.000"
+    # So just grab chars before T and converts to seconds after epoch
+    # then mulitply by 1000000 and convert to int.
+    date_time = datetime.strptime(date_str.split('T')[0], '%Y-%m-%d')
+    if start_of_day:
+        date_time = date_time.replace(hour=0, minute=0, second=0, microsecond=0)
+    else:
+        date_time = date_time.replace(hour=23, minute=59, second=59, microsecond=0)
+    return int(date_time.timestamp() * 1000000)
+
+
+def search_date_range(es_body,
+                      date_pubmed_modified: Optional[List[str]] = None,
+                      date_pubmed_arrive: Optional[List[str]] = None,
+                      date_published: Optional[List[str]] = None):
+    # date_pubmed_X is split to just get the date and remove the time element
+    # as elastic search does a tr comparison and if the end date has the time
+    # element stil in it fails even if the date bits match as the string is longer.
+    if "must" not in es_body["query"]["bool"]["filter"]["bool"]:
+        es_body["query"]["bool"]["filter"]["bool"]["must"] = []
+    if date_pubmed_modified:
+        es_body["query"]["bool"]["filter"]["bool"]["must"].append(
+            {
+                "range": {
+                    "date_last_modified_in_pubmed": {
+                        "gte": date_pubmed_modified[0].split('T')[0],
+                        "lte": date_pubmed_modified[1].split('T')[0]
+                    }
+                }
+            })
+    if date_pubmed_arrive:
+        es_body["query"]["bool"]["filter"]["bool"]["must"].append(
+            {
+                "range": {
+                    "date_arrived_in_pubmed": {
+                        "gte": date_pubmed_arrive[0].split('T')[0],
+                        "lte": date_pubmed_arrive[1].split('T')[0]
+                    }
+                }
+            })
+    if date_published:
+        try:
+            start = date_str_to_micro_seconds(date_published[0], True)
+            end = date_str_to_micro_seconds(date_published[1], False)
+        except Exception as e:
+            logger.error(f"Exception in conversion {e} for start={date_published[0]} and end={date_published[1]}")
+        logger.debug(f"Search date_published: start={start}, end={end}")
+        es_body["query"]["bool"]["filter"]["bool"]["must"].append(
+            {
+                "bool": {
+                    "should": [
+                        {
+                            "range": {
+                                "date_published_end": {
+                                    "gte": start,
+                                    "lte": end
+                                }
+                            }
+                        },
+                        {
+                            "range": {
+                                "date_published_start": {
```

Comment:
I don't know if we need to take into account both date published start and date published end.  Even if we do I think these would need to be in separate clauses.  I would check with Ceri on this but for the sorting story we are only concerned with date published start.

Reply:
For filtering we need to check both I think. Actually I may need another range too.
 There are 3 cases where we want to find papers
1) pub_start is between dates selected,
2) pub_end is between dates selected
3) pub_start > start date selected AND pub_end < end date selected.